src/Umbraco.Web/Services/IconService.cs

```diff
@@ -69,19 +74,58 @@ private IconModel CreateIconModel(string iconName, string iconPath)
         {
             try
             {
-                var svgContent = System.IO.File.ReadAllText(iconPath);
-                var svg = new IconModel
+                return new IconModel
                 {
                     Name = iconName,
-                    SvgString = svgContent
+                    SvgString = System.IO.File.ReadAllText(iconPath)
                 };
-
-                return svg;
             }
             catch
             {
                 return null;
             }
         }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <returns></returns>
+        private IEnumerable<FileInfo> GetAllIconNames()
+        {
+            // add icons from plugins
+            var appPlugins = new DirectoryInfo(IOHelper.MapPath(SystemDirectories.AppPlugins));
+            var pluginIcons = new List<FileInfo>();
+
+            // iterate sub directories of app plugins
+            foreach (var dir in appPlugins.EnumerateDirectories())
+            {
+                var iconPath = dir.FullName + SystemDirectories.AppPluginIcons;
+                if (Directory.Exists(iconPath))
+                {
+                    var dirIcons = new DirectoryInfo(iconPath).EnumerateFiles("*.svg", SearchOption.TopDirectoryOnly)
+                        .Where(x => !pluginIcons.Any(i => i.Name == x.Name));
+                    pluginIcons.AddRange(dirIcons);
+                }
+            }
+
+            // add icons from IconsPath if not already added from plugins
+            var directory = new DirectoryInfo(IOHelper.MapPath($"{_globalSettings.IconsPath}/"));
+            var iconNames = directory.GetFiles("*.svg")
+                .Where(x => pluginIcons.Any(i => i.Name == x.Name) == false)
+                .ToList();
+
+            iconNames.AddRange(pluginIcons);
+
+            return iconNames;
+        }
+
+        private IReadOnlyDictionary<string, string> GetIconDictionary() => _cache.GetCacheItem(
+            "Umbraco.Web.Services.IconService::IconDictionary",
```

Comment:
Better to not hard code these because these namespaces, etc... will change (i.e. netcore). Use:

$"{typeof(IconService).FullName}.{nameof(GetIconDictionary)}"